On E0277 be clearer about implicit `Sized` bounds on type params and assoc types

estebank opened the pull request:
```
error[E0277]: the size for values of type `[i32]` cannot be known at compilation time
   --> f100.rs:2:33
    |
2   |     let _ = std::mem::size_of::<[i32]>();
    |                                 ^^^^^ doesn't have a size known at compile-time
    |
    = help: the trait `Sized` is not implemented for `[i32]`
note: required by an implicit `Sized` bound in `std::mem::size_of`
   --> /home/gh-estebank/rust/library/core/src/mem/mod.rs:312:22
    |
312 | pub const fn size_of<T>() -> usize {
    |                      ^ required by the implicit `Sized` requirement on this type parameter in `size_of`
```

Fix #120178.

estebank committed:
On E0277 be clearer about implicit `Sized` bounds on type params and assoc types

```
error[E0277]: the size for values of type `[i32]` cannot be known at compilation time
   --> f100.rs:2:33
    |
2   |     let _ = std::mem::size_of::<[i32]>();
    |                                 ^^^^^ doesn't have a size known at compile-time
    |
    = help: the trait `Sized` is not implemented for `[i32]`
note: required by an implicit `Sized` bound in `std::mem::size_of`
   --> /home/gh-estebank/rust/library/core/src/mem/mod.rs:312:22
    |
312 | pub const fn size_of<T>() -> usize {
    |                      ^ required by the implicit `Sized` requirement on this bound in `size_of`
```

Fix #120178.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-16`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/7646486528/job/20835391788) [(plain)](https://github.com/rust-lang/rust/commit/0421608954a05ca123614ffff615240975c373d3/checks/20835391788/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Built container sha256:9c3c93a371e5aed5c18185b24f130d95d5140dbd72a9b325e7b6b49e521a4faa
Looks like docker image is the same as before, not uploading
https://ci-caches.rust-lang.org/docker/7ebc15c01a233894034d277c8cce4e949f4e7791f66b4727c8fb6e058a0b8171d6152e1441d677cef0653843ceeee469c097b8699b2bb74249e674f6aa1a8813
sha256:9c3c93a371e5aed5c18185b24f130d95d5140dbd72a9b325e7b6b49e521a4faa
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-16]
##[group]Clock drift check
  local time: Wed Jan 24 21:31:50 UTC 2024
  network time: Wed, 24 Jan 2024 21:31:50 GMT
  network time: Wed, 24 Jan 2024 21:31:50 GMT
##[endgroup]
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-16', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'build.optimized-compiler-builtins', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-16/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
failures:

---- [incremental] tests/incremental/hashes/trait_defs.rs stdout ----

error in revision `cfail2`: test compilation failed although it shouldn't!
status: exit status: 1
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/incremental/hashes/trait_defs.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--cfg" "cfail2" "-C" "incremental=/checkout/obj/build/x86_64-unknown-linux-gnu/test/incremental/hashes/trait_defs/trait_defs.inc" "-Z" "incremental-verify-ich" "-O" "--error-format" "json" "--json" "future-incompat" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/incremental/hashes/trait_defs" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/incremental/hashes/trait_defs/auxiliary" "-Z" "query-dep-graph" "-O" "-Zincremental-ignore-spans"
--- stderr -------------------------------
--- stderr -------------------------------
error: `predicates_of(TraitAddBuiltinBoundToMethodTypeParameter::method)` should be dirty but is not
   |
   |
LL |     fn method<T: Sized>();
Build completed unsuccessfully in 0:13:33

error: aborting due to 1 previous error
------------------------------------------
```

</details>

estebank commented:
This change *as is* affects incremental compilation, making the change `fn foo<T>() {}` to `fn foo<T: Sized>() {}` now be picked up as a change. Haven't looked at the incr compilation machinery in a while, so I'm not sure how long it might take me to fix that.

estebank committed:
Change incr comp test when adding explicit `Sized` bound

Given the previous change to add implicit `Sized` bounds only if there
isn't already an explicit `Sized` bound, now the incr comp machinery
doesn't consider adding the explicit bound as being dirty, as long as
`-Zincremental-ignore-spans` is set.

bors commented:
:umbrella: The latest upstream changes (presumably #120335) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

fmease commented:
I will review this tomorrow together with your other PR :)

r? fmease

fmease commented on compiler/rustc_hir_analysis/src/astconv/bounds.rs:
```diff
@@ -82,8 +90,8 @@ impl<'tcx> dyn AstConv<'tcx> + '_ {
+            // There was in fact a `?Sized`, `!Sized` or `Sized` bound;
             // we don't need to do anything.
         } else if sized_def_id.is_some() {
             // There was no `?Sized` or `!Sized` bound;
```
I guess this could be updated, too. I know those comments are super repetitive and obvious. Proceed however you prefer.

estebank replied:
Done

fmease commented on tests/ui/associated-types/issue-20005.stderr:
```diff
@@ -4,11 +4,11 @@ error[E0277]: the size for values of type `Self` cannot be known at compilation
    |
 LL | trait From<Src> {
-   |            ^^^ required by this bound in `From`
+   |            ^^^ required by the implicit `Sized` requirement on this bound in `From`
```
Could you rephrase this slightly? “requirement on this bound” isn't correct since `Src` isn't a bound but a generic parameter and a bound can't “have” a requirement (it *is* a requirement) and there can't be anything “on” a bound. Of course, this is preexisting but now it's slightly worse. I would go with something like this maybe:

```suggestion
   |            ^^^ required by the implicit `Sized` requirement on this type parameter in `From`
```

estebank replied:
Done.

estebank committed:
review comment: change wording

fmease reviewed (approved):
Thanks a lot!

fmease commented:
@bors r+ rollup error path

bors commented:
:pushpin: Commit 43fdb12b49e213a5241c167230773a685fd97eea has been approved by `fmease`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=fmease 43fdb12b49e213a5241c167230773a685fd97eea -->
<!-- homu: {"type":"Approved","sha":"43fdb12b49e213a5241c167230773a685fd97eea","approver":"fmease","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:umbrella: The latest upstream changes (presumably #120346) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:pushpin: Commit 8b0ab54ffefd8855a9030f2c08463f041d0e9770 has been approved by `fmease`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=fmease 8b0ab54ffefd8855a9030f2c08463f041d0e9770 -->
<!-- homu: {"type":"Approved","sha":"8b0ab54ffefd8855a9030f2c08463f041d0e9770","approver":"fmease","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:evergreen_tree: The tree is currently [closed](https://github.com/rust-lang/rust/pull/120447#issuecomment-1920036755) for pull requests below priority 100. This pull request will be tested once the tree is reopened.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-16`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/7735865103/job/21092236122) [(plain)](https://github.com/rust-lang/rust/commit/8b0ab54ffefd8855a9030f2c08463f041d0e9770/checks/21092236122/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#12 writing image sha256:ee1201a75980e6924e914aafef11979d0195b649a42fc677a1cc962b2a65b343 done
#12 naming to docker.io/library/rust-ci done
#12 DONE 10.0s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-16]
##[group]Clock drift check
  local time: Thu Feb  1 21:43:52 UTC 2024
  network time: Thu, 01 Feb 2024 21:43:52 GMT
  network time: Thu, 01 Feb 2024 21:43:52 GMT
##[endgroup]
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-16', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'build.optimized-compiler-builtins', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-16/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
failures:

---- [incremental] tests/incremental/hashes/trait_defs.rs stdout ----

error in revision `cfail5`: test compilation failed although it shouldn't!
status: exit status: 1
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/incremental/hashes/trait_defs.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--cfg" "cfail5" "-C" "incremental=/checkout/obj/build/x86_64-unknown-linux-gnu/test/incremental/hashes/trait_defs/trait_defs.inc" "-Z" "incremental-verify-ich" "-O" "--error-format" "json" "--json" "future-incompat" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/incremental/hashes/trait_defs" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/incremental/hashes/trait_defs/auxiliary" "-Z" "query-dep-graph" "-O"
--- stderr -------------------------------
--- stderr -------------------------------
error: dep-node label `hir_owner_nodes` not recognized
   |
Build completed unsuccessfully in 0:13:31
Build completed unsuccessfully in 0:13:31
LL | #[rustc_clean(except="hir_owner_nodes", cfg="cfail5")]

error: aborting due to 1 previous error
------------------------------------------

```

</details>

estebank committed:
fix rebase

estebank commented:
@bors r=fmease

bors commented:
:pushpin: Commit 0df6dfdc78624282b4193fc2290ba79d26ba6db3 has been approved by `fmease`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=fmease 0df6dfdc78624282b4193fc2290ba79d26ba6db3 -->
<!-- homu: {"type":"Approved","sha":"0df6dfdc78624282b4193fc2290ba79d26ba6db3","approver":"fmease","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
